google/cloud/bigquery/ipython_magics/line_arg_parser/lexer.py

```diff
@@ -0,0 +1,88 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from collections import namedtuple
+from collections import OrderedDict
+
+
+Token = namedtuple("Token", ("type_", "lexeme", "pos"))
+
+# Token definitions.
+# @enum.unique()
+# class TokenType(str, enum.Enum):
+#     def _generate_next_value_(name, start, count, last_values):
+#         return name
+#
+#     COMMA = enum.auto()
+#     DOLLAR = enum.auto()
+
+token_types = OrderedDict(
+    DEST_VAR=r"(?P<DEST_VAR>(?:^\s*)[^\d\W]\w*)",  # a Python ID at the very beginning
+    PY_STRING=r"(?P<PY_STRING>(?:{})|(?:{}))".format(
+        r"'(?:[^'\\]|\.)*'", r'"(?:[^"\\]|\.)*"'  # single and double quoted strings
+    ),
+    OPT_VAL=r"(?P<OPT_VAL>[^_\d\W](?:\w|\.)+)",  # currently the only allowed value is project ID
+    PY_BOOL=r"(?P<PY_BOOL>True|False)",
+    DOLLAR_PY_ID=r"(?P<DOLLAR_PY_ID>\$[^\d\W]\w*)",
+    PY_ID=r"(?P<PY_ID>[^\d\W]\w*)",
+    PY_NUMBER=r"(?P<PY_NUMBER>-?[1-9]\d*(?:\.\d+)?)",  # TODO: into or flowats, exclude exponent notation and other stuff)
+    PARAMS_OPT_SPEC=r"(?P<PARAMS_OPT_SPEC>--params)",
+    OPTION_SPEC=r"(?P<OPTION_SPEC>--\w+)",
+    SQUOTE=r"(?P<SQUOTE>')",
+    DQUOTE=r'(?P<DQUOTE>")',
+    COLON=r"(?P<COLON>:)",
+    COMMA=r"(?P<COMMA>,)",
+    LCURL=r"(?P<LCURL>\{)",
+    RCURL=r"(?P<RCURL>})",
+    LSQUARE=r"(?P<LSQUARE>\[)",
+    RSQUARE=r"(?P<RSQUARE>])",
+    LPAREN=r"(?P<LPAREN>\()",
+    RPAREN=r"(?P<RPAREN>\))",
+    WS=r"(?P<WS>\s+)",
+    EOL=r"(?P<EOL>$)",
+    UNKNOWN=r"(?P<UNKNOWN>\S+)",  # anything not a whitspace or matched by something else
+)
+
+
+class Lexer(object):
+    """Lexical analyzer for tokenizing the cell magic input line."""
+
+    _grand_pattern = re.compile("|".join(token_types.values()))
+
+    def __init__(self, input_text):
+        self._text = input_text
+
+    def __iter__(self):
+        # TOOD: we can "cheat" and split options and positional args ... str.partition()
+        # # explain why... problems with overlapping DEST_VAR and PY_ID and OPT_VAL
+        #
+        # everything before "--" are positional args that we parse separately,
+        # and we parse the rest with what we have
+        # this avoids the hack for DEST_VAR
+        #
+        # of course, in the second group (parsing --options), adjust the position offset
+        #
+        # This means having two grand patterns? Makes sense!
+        #
+        # And rename this to iter_tokens()?
+
+        # BUT !!!! do this after the first working version, write a few Visitors
```

Comment:
TL; DR - we need to add lexer states.

Since `scanner` does not seem to expose a way to examine/set states, we need to implement them manually and group tokens accordingly, so that in each Lexer state there's only a single correct order of token definitions, i.e. no ambiguity.